Add VPC site-to-site tutorial

helge000 opened the pull request:
I have read and understood the Contributor's Certificate of Origin available at the end of https://raw.githubusercontent.com/hetzneronline/community-content/master/tutorial-template.md and I hereby certify that I meet the contribution criteria described in it.
Signed-off-by: DANIEL HELGENBERGER <daniel.helgenberger@m-box.de>

svenja11 commented:
Thank you for your contribution @helge000! I read through your tutorial and I have a few comments.

* Please add a description in line 7.
* The letter `f` in `pfSense` is missing in one of your images. In ` vpc-network.png`, it says `pSense`.
* Please do not use actual IP addresses in your tutorial. You can check out [tutorial-template.md](https://github.com/hetzneronline/community-content/blob/master/tutorial-template.md#introduction) for a couple of example IPs, such as `<203.0.113.1>` and ` <2001:db8:1234::1> `. I have updated the screenshots of the Cloud Console and the VNC console on Cloud Console for you. Please check if the other images also need to be updated.
* Please change the sentences in lines 128 and 289.

---------

Changes I made:

* I swapped the images in steps 5.1 and 5.2
* In step 5.3, I have added a link to the Hetzner getting started "[Connecting to a Server via its private IP](https://docs.hetzner.com/cloud/servers/getting-started/connecting-via-private-ip)"

svenja11 commented:
@helge000  Do you still plan to work on this tutorial?

helge000 commented:
@svenja11 - thanks for the reminder, and yes - indeed. Had a lot of other things to do lately but will fix this shortly. Cheers!

helge000 commented:
@svenja11 , thanks for fixing all my spelling stuff. I updated the public IP on-prem; but left the one with Hetzner. Reasoning:
- The shown IP belongs to Hetzner's RIPE range anyway;
- I need to make it really clear what address I am referring to in the examples. It helps to make them as different as possible.

> Please change the sentences in lines 128 and 289.

I did not find the sentence in line 128 you are referring to. Please directly open a review at this line. The other one is wiced :)

please note:
- rebased to current HEAD
- I will squash once we're done

svenja11 commented on tutorials/vpc-with-wireguard-pfsense/configs/aliases-config-vpc-gw.example.com.xml:
```diff
@@ -0,0 +1,45 @@
+		<name>full_net</name>
+		<type>network</type>
+		<address>vpc_net on_prem_net</address>
+		<descr><![CDATA[Exampe Corp Internal segment]]></descr>
```
````suggestion
		<descr><![CDATA[Example Corp Internal segment]]></descr>
```

svenja11 commented on tutorials/vpc-with-wireguard-pfsense/configs/aliases-config-vpc-gw.example.com.xml:
```diff
@@ -0,0 +1,45 @@
+		<name>on_prem_net</name>
+		<type>network</type>
+		<address>10.0.0.0/17</address>
+		<descr><![CDATA[Exampe Corp On Prem segment]]></descr>
```
```suggestion
		<descr><![CDATA[Example Corp On Prem segment]]></descr>
```

svenja11 commented on tutorials/vpc-with-wireguard-pfsense/configs/aliases-config-vpc-gw.example.com.xml:
```diff
@@ -0,0 +1,45 @@
+		<name>vpc_net</name>
+		<type>network</type>
+		<address>10.0.128.0/17</address>
+		<descr><![CDATA[Exampe Corp VPC segment]]></descr>
```
```suggestion
		<descr><![CDATA[Example Corp VPC segment]]></descr>
```

svenja11 commented on tutorials/vpc-with-wireguard-pfsense/01.en.md:
```diff
@@ -0,0 +1,382 @@
+
+**Please note** that there has been [some controversy](https://www.netgate.com/blog/painful-lessons-learned-in-security-and-community) around Netgate's WireGuard implementation in the past. The current kmod-plugin for FreeBSD (what the [pfSense](https://github.com/pfsense/FreeBSD-ports/tree/devel/net/pfSense-pkg-WireGuard) uses in turn) is developed by the [same guy](https://git.zx2c4.com/wiregua ...
+
+### Overview
+
+1. Install the wireguard package on both pfSense instances
+2. Set up the server on both instances
+3. Add firewall rules to allow traffic on the wireguard interface
+4. Set up a peer (client) witch on each instance "dials" into the other pfSense
+5. Add interfaces and gateways - this should bring up the tunnel
+6. Set up static routes
+7. Disable NAT translation
+8. Deploy a test VM, setup the default route, install `iperf3` and test the tunnel's performance
+
+Follow the [official guide](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html); you will see the same addresses there. It is recommended to use an address segment outside our *Example Corp* network. Further, it is recommended to set up a [Pre Shared Key](https://docs.netgate.com/pfsense/en/latest/vpn/wireguard/settings.html#wireguard-tunnel-settings).
+
+### Remarks
+
+To clarify the guide from pfSense, make sure to follow all steps on both of your pfSense instances. The following only shows the steps on VPC but they are essentially the same on your on-prem pfSense instance.
+
+- Allow inbound traffic on port `51820/udp` on all of your WAN interfaces.
+  Should you use another port, ensure to follow through with it!
+- Make sure to actually *enable* the WireGuard service in the settings and set *Interface Group Membership* to *Unassigned Tunnels*.
+  This will give you more control over the tunnel rules.<br>
+  ![WireGuard Settings](images/pfsense-wg-settings.png)
+- Remember to set the allowed IP networks in the peer. These are the same on both ends of the tunnel. Further, generate a PSK here and copy it to the peer of the other instance. Since we create a site-to-site connection, it is recommended to have static peer addresses. It is highly recommended to use *publicly resolvable DNS entries here* instead of the IP addresses.</br>
+  ![WireGuard Peer setup](images/pfsense-wg-peer-setup.png)
+- Ensure to assign the interface in `Interfaces` / `Interface Assignments`</br>
+  ![pfsense-wg-if-assign.png](images/pfsense-wg-if-assign.png)
+- and set up the gateway correctly. The gateway is the address of the other instance (see [diagram](https://github.com/hetzneronline/community-content/tree/master/tutorials/vpc-with-wireguard-pfsense/images/vpc-network.png)). **Very important:** Set the [MTU](https://en.wikipedia.org/wiki/Maximum_transmission_unit) on both sides to a smaller value, [it is a good idea](https://keremerkan.net/posts ...
+  ![pfsense-wg-interface.png](images/pfsense-wg-interface.png)
+- When everything is set correctly on both ends, the tunnel should come up with a green handshake icon in the status page and a handshake time. Check if the tunnel is up on both instances.</br>
+  ![pfsense-wg-status-page.png](images/pfsense-wg-status-page.png)
+- **The crucial part** is to add our routes. You have to add the OnPrem segment via the OnPrem WireGuard Gateway on the VPC pfSense instance; and the route to the VPC segment via the VPC WireGuard Gateway on the OnPrem instance. Below the entry for our VPC pfSense instance</br>
+  ![VPC static routes](images/pfsense-wg-route.png)
```
"Overview" in line 214 and "Remarks" in line 227 kind of link together. I would suggest to add your comments directly to the individual steps for a better reading flow. For example:

```suggestion
Install the wireguard package on both pfSense instances.

In your pfSense web interface, go to `System` / `Package Manager` and select the "Available Packages" tab. Search for "WireGuard" and click on `+ Install`. After it is installed, follow the [official guide](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html); you will see the same addresses there. It is recommended to use an address segment outside our *Example Corp* network. Further, it is recommended to set up a [Pre Shared Key](https://docs.netgate.com/pfsense/en/latest/vpn/wireguard/settings.html#wireguard-tunnel-settings).

To clarify the guide from pfSense, make sure to follow all steps on both of your pfSense instances. The following only shows the steps on VPC but they are essentially the same on your on-prem pfSense instance.

1. [Add firewall rules](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#firewall-rules) to allow traffic on the wireguard interface
   
   Allow inbound traffic on port `51820/udp` on all of your WAN interfaces.
   Should you use another port, ensure to follow through with it!

2. [Enable WireGuard](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#wireguard-configuration) under `VPN` / `WireGuard` / Settings
   
   Make sure to actually *enable* the WireGuard service in the settings and set *Interface Group Membership* to *Only Unassigned Tunnels*. This will give you more control over the tunnel rules.
   
   ![WireGuard Settings](images/pfsense-wg-settings.png)

3. [Add a tunnel](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#tunnel-configuration)

4. [Set up a peer](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#peer-configuration) (client) which "dials" into the other pfSense instance
   
   Remember to set the allowed IP networks in the peer. These are the same on both ends of the tunnel. Further, generate a PSK here and copy it to the peer of the other instance. Since we create a site-to-site connection, it is recommended to have static peer addresses. It is highly recommended to use *publicly resolvable DNS entries here* instead of the IP addresses.

   ![WireGuard Peer setup](images/pfsense-wg-peer-setup.png)

5. [Add interfaces and gateways](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#assign-interface) - this should bring up the tunnel
   
   Ensure to assign the interface in `Interfaces` / `Interface Assignments`
   
   ![pfsense-wg-if-assign.png](images/pfsense-wg-if-assign.png)

   and set up the gateway correctly. The gateway is the address of the other instance (see [diagram](https://github.com/hetzneronline/community-content/tree/master/tutorials/vpc-with-wireguard-pfsense/images/vpc-network.png)). **Very important:** Set the [MTU](https://en.wikipedia.org/wiki/Maximum_transmission_unit) on both sides to a smaller value, [it is a good idea](https://keremerkan.net/posts/wireguard-mtu-fixes/) to start out as low as **1300** bytes. Set the same MTU on both ends of the tunnel.
   
   ![pfsense-wg-interface.png](images/pfsense-wg-interface.png)

6. [Set up static routes](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html#routing)
   
   **The crucial part** is to add our routes. You have to add the OnPrem segment via the OnPrem WireGuard Gateway on the VPC pfSense instance; and the route to the VPC segment via the VPC WireGuard Gateway on the OnPrem instance. Below, the entry for our VPC pfSense instance:

   ![VPC static routes](images/pfsense-wg-route.png)

7. Disable NAT translation

When everything is set correctly on both ends, the tunnel should come up with a green handshake icon in the status page and a handshake time. Check if the tunnel is up on both instances.

![pfsense-wg-status-page.png](images/pfsense-wg-status-page.png)

Deploy a test VM, setup the default route, install `iperf3` and test the tunnel's performance.
```

helge000 replied:
Hi @svenja11 , the intention for `Remarks` is in reference to the Netgate WireGuard guide itself. I'd like to keep this separated therefore. However, my intention is not obvious I guess. 
Further, the `Overview` section is meant as an overview for the *whole* tutorial (it is actually part of the template, iirc). 

That said, I'd rather keep it separated. Do you have any suggestions to make this more clear?

svenja11 replied:
Hi @helge000, I actually did understand that you wanted `Remarks` to be in reference to the Netgate WireGuard guide itself.

But I guess I did get the purpose of `Overview` wrong. Maybe because it is part of "Step 4". I thought, in this overview, you were just listing the different steps that the reader is supposed to follow in the Netgate WireGuard guide at [docs.netgate.com](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html). So this is how I thought it is structured:

> **Step 4 - Setting up WireGuard site-to-site connection**
> 
> Overview
> *- List of Netgate guide steps -*
> Follow the [official guide](https://docs.netgate.com/pfsense/en/latest/recipes/wireguard-s2s.html) ...
>
> Remarks
> *- Notes about the Netgate guide steps -*

So I thought it would make sense to combine it like this:

> **Step 4 - Setting up WireGuard site-to-site connection**
> 
> *- List of Netgate guide steps with your personal notes -*

---------------------

Usually, the `Introduction` at the top is used to give a brief overview over the different steps in the tutorial. So maybe that's why I didn't get that the overview in "Step 4" was also for the *whole* tutorial. If you want to leave it the way it is, just let me know and I will mark it as "ready".

helge000 commented:
> I also made a few suggestions. In lines 7, 14, and 28 you wrote "Exampe" instead of "Example". I noticed the same typo in the screenshot pfsense-aliases.png. Can you update the screenshot?

@svenja11 - 210417b fixes the screenshot - sorry for so many example typos ;)

helge000 commented:
@svenja11, thanks for your help.

>  If you want to leave it the way it is, just let me know and I will mark it as "ready".

Yes, indeed. I rebased and squashed this PR.
